[ART] WAL serialization, automatic checkpointing, decoupling catalog and storage, index names

taniabogatsch opened the pull request:
This is an involved ART PR that mainly covers four topics. All these topics depend on each other, so I decided to bundle them in on PR—general related issues: https://github.com/duckdblabs/duckdb-internal/issues/199, https://github.com/duckdblabs/duckdb-internal/issues/102.

1. Serializing small indexes directly to the WAL. Before this PR, we would rebuild the ART when replaying the WAL. This led to hard-to-fix index binder issues and specialized, non-performant index creation code.
2. Index changes are now accounted for when estimating the automatic checkpoint threshold.
3. Decoupling index catalog entries and their physical storage by removing the `index` field from the `IndexCatalogEntry`. This is a step towards pluggable indexes and also towards custom catalogs.
4. Primary keys and foreign keys now have names.

### WAL Serialization

We reuse most of our existing serialization code for serializing an ART index to the WAL. However, we do not need to write the allocator data onto partial blocks, as we are directly writing the bytes from their respective buffers. Therefore, we pass a flag to the existing index serialization functions. Then, we only return all `data_ptr_t`s of relevant buffers and their `allocation_size`.  Related fixed issues: https://github.com/duckdblabs/duckdb-internal/issues/345, https://github.com/duckdblabs/duckdb-internal/issues/374

**NOTE**: I changed the index serialization to work with the WAL and checkpoint managers. Before this PR, we used the metadata manager for serializing non-block index data, which is not accessible from the WAL. Now, we use the `Binary(De)Serializer` directly. This also means adding custom serialization code to the respective JSON files for new storage information structs. To guarantee backward compatibility, I kept some of the old deserialization code.

**NOTE**: THIS BREAKS FORWARD STORAGE COMPATIBILITY. Unifying the index serialization code to use the `Binary(De)Serializer` avoids two separate implementations (WAL and checkpoint manager). If we do not want this, the implementation becomes more cumbersome, and I will implement two different code paths (@Maxxen?).


### Account Indexes for Automatic Checkpoints

When estimating the size of all changes in a transaction, we now consider an estimated index size. This enables indexes to trigger automatic checkpoints (https://github.com/duckdb/duckdb/issues/6273).


### Decoupling the Index Catalog Entry from the Physical Index Storage

I removed the physical index from the index catalog entry. Also, I added names to primary and foreign keys. Physical index objects now also contain their names. These names are the connection between a catalog entry and an index object. We need this connection in various places where we directly used the linked physical object in the catalog entry previously.


Additionally, I removed the `IndexType` enum, which is now simply a string. This allows adding arbitrary new index types through extensions without changing a duckdb enum. Related issue: https://github.com/duckdblabs/duckdb-internal/issues/321


### Questions
`src/include/duckdb/storage/serialization/create_info.json`: I changed a type here and removed a field. Is this allowed?


### Follow-Up PRs

We now use an `initial_art_size` variable in the `DuckIndexEntry`. I plan to remove this variable again in a follow-up PR that dives more into the unique names of indexes. In this follow-up PR, I plan to add a function to traverse the catalog and execute a callback function on all existing indexes (via `DuckTableEntry`). Two usages for this callback function are (1) checking for unique name conflicts with primary and possibly foreign keys during index creation and (2) retrieving a matching index from the undo buffer during WAL size estimation.

At the moment, conflicting names between `CREATE INDEX` and primary/foreign keys are not problematic, as we always deserialize primary/unique/foreign indexes before separately created indexes, and because it is not yet possible to set custom names for these constraints.

taniabogatsch committed:
WIP index fixes and improvements

taniabogatsch committed:
compiles

taniabogatsch committed:
fixes

taniabogatsch committed:
tidy fixes

taniabogatsch committed:
backwards compatibility

taniabogatsch committed:
backwards compatibility fixes

taniabogatsch committed:
Merge branch 'feature' into art-wal

taniabogatsch committed:
minor fixes

taniabogatsch committed:
Windows include fixes

taniabogatsch committed:
more include clean-up

taniabogatsch committed:
directly (de)serializing to/from the WAL

taniabogatsch committed:
auto checkpoint after index creation and insertions

taniabogatsch committed:
Merge branch 'feature' into art-wal

taniabogatsch committed:
Merge branch 'feature' into art-wal

taniabogatsch committed:
serializing bytes to the WAL directly

taniabogatsch committed:
tidy checks

taniabogatsch committed:
more missing exclude and build fixes

taniabogatsch committed:
more fixes

taniabogatsch committed:
not changing FixedSizeBuffer during WAL serialization

taniabogatsch committed:
track initial index size

taniabogatsch committed:
Merge branch 'feature' into art-wal

taniabogatsch committed:
tidying up

taniabogatsch committed:
regenerate serialization

taniabogatsch committed:
missing include

taniabogatsch committed:
Merge branch 'feature' into art-wal

pdet commented on src/execution/index/art/art.cpp:
```diff
@@ -33,15 +30,16 @@ struct ARTIndexScanState : public IndexScanState {
-         const BlockPointer &pointer)
-    : Index(db, IndexType::ART, table_io_manager, column_ids, unbound_expressions, constraint_type),
+         const IndexStorageInfo &index_storage_info)
+    : Index(name, "ART", index_constraint_type, column_ids, table_io_manager, unbound_expressions, db),
```
So, the index type here is changed to allow for pluggable indexes. 
I'm wondering if it makes sense to have a dynamic enum here (i.e., a static enum with an extra map for indexes created in extensions) to avoid error-prone string comparisons switcharoo.
I think these switcharoos will have to happen in the core, since theoretically, you can have multiple extensions with different index structures that must co-exist. But maybe we can have a bigger discussion on the pluggable indexes strategy later. cc @Maxxen

pdet commented on src/execution/index/art/art.cpp:
```diff
@@ -58,8 +56,23 @@ ART::ART(const vector<column_t> &column_ids, TableIOManager &table_io_manager,
+			InitAllocators(index_storage_info);
+
+		} else {
+			// STABLE STORAGE NOTE: this code path is necessary to read older duckdb files
+			D_ASSERT(index_storage_info.properties.find("block_id") != index_storage_info.properties.end());
+			D_ASSERT(index_storage_info.properties.find("offset") != index_storage_info.properties.end());
+
+			auto block_id = index_storage_info.properties.find("block_id")->second;
+			auto offset = index_storage_info.properties.find("offset")->second;
+			BlockPointer block_pointer(block_id.GetValue<block_id_t>(), offset.GetValue<uint32_t>());
+			if (block_pointer.IsValid()) {
+				Deserialize(block_pointer);
+			}
```
Maybe move it to a module?

pdet commented on src/execution/index/fixed_size_buffer.cpp:
```diff
@@ -195,6 +192,15 @@ uint32_t FixedSizeBuffer::GetOffset(const idx_t bitmask_count) {
+void FixedSizeBuffer::SetAllocationSize(const idx_t available_segments, const idx_t segment_size,
+                                        const idx_t bitmask_offset) {
+
+	if (dirty) {
```
dirty here means invalid?

pdet commented on src/parser/parsed_data/create_index_info.cpp:
```diff
@@ -6,10 +6,15 @@ unique_ptr<CreateInfo> CreateIndexInfo::Copy() const {
 	auto result = make_uniq<CreateIndexInfo>();
 	CopyProperties(*result);
 
-	result->index_type = index_type;
+	result->table = table;
 	result->index_name = index_name;
+
+	result->options = options;
+
+	result->index_type = index_type;
 	result->constraint_type = constraint_type;
-	result->table = table;
+	result->column_ids = column_ids;
+
```
Maybe we use a constructor?

pdet commented on src/parser/transform/statement/transform_create_index.cpp:
```diff
@@ -42,18 +41,9 @@ unique_ptr<CreateStatement> Transformer::TransformCreateIndex(duckdb_libpgquery:
-	}
-
-	info->index_type_name = index_type_name;
+	info->index_type = StringUtil::Upper(string(stmt.accessMethod));
```
this is the type of thing that worries me about strings as well

pdet commented on src/planner/expression_binder/index_binder.cpp:
```diff
@@ -20,28 +20,6 @@ BindResult IndexBinder::BindExpression(unique_ptr<ParsedExpression> &expr_ptr, i
 	case ExpressionClass::SUBQUERY:
 		return BindResult("cannot use subquery in index expressions");
 	case ExpressionClass::COLUMN_REF: {
```
maybe we move column_ref to default?

pdet commented on src/storage/checkpoint/table_data_writer.cpp:
```diff
@@ -66,13 +66,21 @@ void SingleFileTableDataWriter::FinalizeTable(TableStatistics &&global_stats, Da
+	serializer.WriteProperty(101, "table_pointer", pointer);
+	serializer.WriteProperty(102, "total_rows", total_rows);
+
+	// STABLE STORAGE NOTE: We write an index pointers vector with exactly one invalid index pointer with a valid
+	// offset, because when reading the file, we don't know if we're reading an older storage version (with
+	// BlockPointers) or not. The combination of invalid block ID and valid offset allows us to distinguish
+	vector<BlockPointer> index_pointers;
+	index_pointers.emplace_back(INVALID_BLOCK, 42);
+	serializer.WriteProperty(103, "index_pointers", index_pointers);
+
+	auto index_storage_infos = info->indexes.GetStorageInfos();
+	serializer.WriteList(104, "index_storage_infos", index_storage_infos.size(),
+	                     [&](Serializer::List &list, idx_t i) { list.WriteElement(index_storage_infos[i]); });
```
move it to a modular function as well?

These stable storage codes are small but also look a bit dirty, maybe they should be moved somewhere (static functions in a Index_storage with the comments?

pdet commented on src/storage/checkpoint_manager.cpp:
```diff
@@ -435,19 +436,30 @@ void CheckpointReader::ReadIndex(ClientContext &context, Deserializer &deseriali
+	IndexStorageInfo index_storage_info;
+	if (root_block_pointer.IsValid()) {
+		// STABLE STORAGE NOTE: this code path is necessary to read older duckdb files
+		index_storage_info.name = "ART_index";
```
This won't cause any conflicts if you have multiple indexes being read?

pdet commented on src/storage/wal_replay.cpp:
```diff
@@ -403,47 +404,86 @@ void ReplayState::ReplayDropTableMacro(BinaryDeserializer &deserializer) {
+
+		// read the data into buffer handles and convert them to blocks on disk
+		// then, update the block pointer
+		for (idx_t j = 0; j < data_info.buffer_block_pointers.size(); j++) {
```
this can be a `auto& buffer_block_ptr`?

taniabogatsch replied:
I need `j` for another equally sized vector that contains the allocation sizes when reading the bytes: `list.ReadElement<bool>(data_ptr, data_info.buffer_allocation_sizes[j]);`

pdet reviewed (commented):
Very exciting work and it is cool to see that pluggable indexes are coming! :-)

I've added some comments, mostly NIPs, besides the index_types becoming strings and adding a lot of code to manage old checkpoints. You can see my comments about them around the code, but we can chat about it in more detail if you wish!

For the pluggable indexes, I think it would be cool to see a mock index being plugged as a test, but it's also a bit out of the scope of this PR.

Otherwise, I'm missing a couple more tests.
1. Can we have a test that loads a DuckDB with an index on an older storage? Different types of indexes, multiple indexes in the same db, and so on.
2. Can we add tests that check indexes with the same name can't be created? including scenarios with camelcases, invalid names, etc..

taniabogatsch committed:
Merge branch 'feature' into art-wal

Maxxen commented:
Re: Index type strings - we need to have some sort of identifier for index types, and a string is probably the most flexible way to do it. 

Ideally I think it makes most sense to have "IndexType" being a separate catalog entry, which an index "instance" can reference. Similarly to how we currently can create custom types, and then have tables/functions with columns/arguments referencing these types. Im not sure how these sort of intra-catalog references are made though, is this where you would use an OID? How does it work now when e.g. a table references an index? If that's just by name (i.e string) then I don't see the issue with index types being strings too.

taniabogatsch commented:
> to have "IndexType" being a separate catalog entry, which an index "instance" can reference

This also sounds like a good solution to me. 

> Im not sure how these sort of intra-catalog references are made though, is this where you would use an OID?

I think I would also have to look into this a bit more. I am fine with keeping the `IndexType` enum for this PR. And then changing directly to a more elaborate solution in a future PR. From this discussion, I think that that's, for now, a good solution, and lets me come up with something more sustainable, as I also get @pdet's arguments against strings.

> How does it work now when e.g. a table references an index? If that's just by name (i.e string) then I don't see the issue with index types being strings too.

It's very 'naive' at the moment (before this PR) in my opinion, physical indexes do not even have names, and we'd only access/change them because they were directly linked to their catalog counterpart. 

I added an initial attempt at naming support on this PR, but this discussion and other comments made me realize that I should improve the index names more, as they must be more robust now since they're the link between physical indexes and catalog entries. Postgres, e.g., treats indexes as relations, in a way.

taniabogatsch committed:
revert back to IndexType for this PR and some other review suggestions

taniabogatsch committed:
move to constructor

taniabogatsch committed:
Merge branch 'feature' into art-wal

taniabogatsch commented:
> For the pluggable indexes, I think it would be cool to see a mock index being plugged as a test, but it's also a bit out of the scope of this PR.

Yes, this is out of the scope of this PR, and I also think that we are not completely there yet.

taniabogatsch committed:
more review feedback

taniabogatsch committed:
Merge branch 'readonly-idx-bug' into art-wal

taniabogatsch committed:
Merge branch 'feature' into art-wal

taniabogatsch committed:
Merge branch 'pk-mem-cleanup' into art-wal

taniabogatsch committed:
more testing of duckdb versions

taniabogatsch committed:
generate unique pk/fk/unique names for older duckdb versions

taniabogatsch committed:
formatting, tidying things up and more modularization

taniabogatsch committed:
add 0-9-1 duckdb storage file

taniabogatsch commented:
Hey @pdet, thanks for your review! I've now implemented all your suggestions. Maybe you can have another look?

> Can we have a test that loads a DuckDB with an index on an older storage? Different types of indexes, multiple indexes in the same db, and so on.

I've now added another file for explicit compatibility with older duckdb file versions: `test/sql/index/art/storage/test_art_duckdb_versions.test`. Indexes are already part of our `test/sql/storage_version/storage_version.test_slow` routine, but testing them more explicitly can't hurt.

> Can we add tests that check indexes with the same name can't be created? including scenarios with camelcases, invalid names, etc..

There already is a test file `test/sql/index/art/storage/test_art_names.test`. Anything specific that you are missing there?

taniabogatsch committed:
add semicolon

taniabogatsch committed:
missing include

taniabogatsch committed:
Merge branch 'feature' into art-wal

taniabogatsch committed:
Merge branch 'feature' into art-wal

Mytherin commented on data/storage/index_0-9-1.db:
In the future can we add support for reading gzipped DuckDB files to the test runner so we can store these files compressed? This matters because these files are small.

Mytherin commented on test/sql/index/art/storage/test_art_names.test:
```diff
@@ -0,0 +1,81 @@
+restart
+
+statement ok
+CREATE INDEX PRIMARY_tbl_0 ON tbl(i);
```
Looks like this should be `PRIMARY_tbl0`

Mytherin commented on test/sql/index/art/storage/test_art_names.test:
```diff
@@ -0,0 +1,81 @@
+DROP INDEX UNIQUE_tbl_1;
+
+statement ok
+DROP INDEX PRIMARY_tbl_0;
```
Looks like this actually removes the internal primary key once renamed - could you check/fix that?

Mytherin commented on src/include/duckdb/storage/table_storage_info.hpp:
```diff
@@ -31,6 +34,45 @@ struct ColumnSegmentInfo {
+	vector<idx_t> buffer_counts;
+
+	//! When serializing to the WAL, this contains data_ptr_t to each pinned buffer, and their allocation size
+	vector<vector<pair<data_ptr_t, idx_t>>> buffers;
```
Could we turn this into a struct?

Mytherin commented on src/include/duckdb/catalog/catalog_entry/index_catalog_entry.hpp:
```diff
@@ -24,21 +22,37 @@ class IndexCatalogEntry : public StandardEntry {
+	case_insensitive_map_t<Value> options;
+
+	//! The index type (ART, B+-tree, Skip-List, ...)
+	IndexType index_type;
```
Could you move this to the `DuckIndexEntry` instead?

taniabogatsch replied:
We also need these fields for (de)serialization in `CreateIndexInfo`. And on second thought, I think these fields apply to many indexes, not specifically the `DuckIndexEntry`, no? Or would I create a separate `CreateDuckIndexInfo`? I didn't see that for any other entry, e.g., there is no `CreateDuckTableInfo`.

Mytherin commented on src/include/duckdb/execution/index/fixed_size_allocator.hpp:
```diff
@@ -75,10 +73,16 @@ class FixedSizeAllocator {
+	vector<pair<data_ptr_t, idx_t>> InitSerializationToWAL();
+	//! Initialize a fixed-size allocator from allocator storage information
+	void Init(const IndexDataInfo &info);
+	//! STABLE STORAGE NOTE: This is for old storage files, to deserializes all metadata
```
Could you turn off caps lock here

Mytherin commented on src/include/duckdb/parser/parsed_data/create_index_info.hpp:
```diff
@@ -8,42 +8,44 @@
+	case_insensitive_map_t<Value> options;
+
+	//! The index type (ART, B+-tree, Skip-List, ...)
+	IndexType index_type;
```
Let's only keep this as a string and remove the `IndexType` for extensibility

Mytherin commented on src/storage/checkpoint_manager.cpp:
```diff
@@ -375,38 +375,39 @@ void CheckpointReader::ReadSequence(ClientContext &context, Deserializer &deseri
+
+	// also, we write a dummy BlockPointer, because when reading the file, we don't know if we're
+	// reading an older storage version (with BlockPointers) or not
+	serializer.WriteProperty(101, "root_block_pointer", BlockPointer());
```
If you want to delete this property we can remove the write, and on the read have a default value (`BlockPointer()`) using `ReadPropertyWithDefault`

Mytherin commented on src/storage/index.cpp:
```diff
@@ -107,4 +119,47 @@ string Index::AppendRowError(DataChunk &input, idx_t index) {
+
+	vector<BlockPointer> index_pointers;
+	index_pointers.emplace_back(INVALID_BLOCK, 42);
+	serializer.WriteProperty(103, "index_pointers", index_pointers);
```
Maybe here we can also not write it and use a read with default

Mytherin commented on src/storage/checkpoint_manager.cpp:
```diff
@@ -519,8 +528,22 @@ void CheckpointReader::ReadTableData(ClientContext &context, Deserializer &deser
 	auto total_rows = deserializer.ReadProperty<idx_t>(102, "total_rows");
+
+	// for current duckdb file versions, this will read a vector with exactly one {-1, 42} index pointer
 	auto index_pointers = deserializer.ReadProperty<vector<BlockPointer>>(103, "index_pointers");
```
Can we use ReadPropertyWithDefault here?

Mytherin commented on src/storage/checkpoint_manager.cpp:
```diff
@@ -519,8 +528,22 @@ void CheckpointReader::ReadTableData(ClientContext &context, Deserializer &deser
 	auto index_pointers = deserializer.ReadProperty<vector<BlockPointer>>(103, "index_pointers");
 
+	if (!index_pointers.empty() && !index_pointers.front().IsValid() && index_pointers.front().offset == 42) {
+		deserializer.ReadList(104, "index_storage_infos", [&](Deserializer::List &list, idx_t i) {
```
This switcheroo might not be necessary if we use `ReadPropertyWithDefault` in both cases, e.g.:

```


Write (old)
103: vector[...]


Write (new)
104: vector[...]


auto old_list = ReadPropertyWithDefault(103, "")
auto new_list = ReadPropertyWithDefault(104, "");

if (old_list.empty()) {
	// 103 was not there, use 104
} else {
	// 104 was not there, use 103
}

```

Mytherin replied:
Can we remove the `deserializer.ReadList` here and use the regular deserialization code?

Mytherin commented on src/storage/index.cpp:
```diff
@@ -107,4 +119,47 @@ string Index::AppendRowError(DataChunk &input, idx_t index) {
+BlockPointer IndexStorage::GetBlockPointer(const IndexStorageInfo &info) {
+
+	D_ASSERT(info.properties.find("block_id") != info.properties.end());
+	D_ASSERT(info.properties.find("offset") != info.properties.end());
```
Maybe instead of having key/value `properties` we can have inheritance, similar to what we have for e.g. `BindData` of functions, allowing us to use regular C++ properties while still allowing extension data to be serialized?

taniabogatsch replied:
I removed this from the PR altogether, as pluggable indexes are not necessarily the scope here. I'll think about inheritance when Max and I start adding the first pluggable index.

Mytherin commented on src/storage/wal_replay.cpp:
```diff
@@ -403,47 +404,86 @@ void ReplayState::ReplayDropTableMacro(BinaryDeserializer &deserializer) {
+	auto &block_manager = single_file_sm.block_manager;
+	auto &buffer_manager = block_manager->buffer_manager;
+
+	deserializer.ReadList(103, "index_storage", [&](Deserializer::List &list, idx_t i) {
```
Do we need to do the `ReadList` here?

taniabogatsch replied:
In `write_ahead_log.cpp`, we write `allocation_size` bytes at the `buffer_ptr` position. The list that we serialize there, in `SerializeIndexToWAL`, contains all ART blocks from which we have to write data.

Here, we need to read these bytes for each list entry. I don't know how to generate code for that. As I can tell the serializer to write the struct, but not to write `allocation_size` bytes at `buffer_ptr`.

Relevant code:
```cpp
	//! Contains all buffer pointers and their allocation size for serializing to the WAL
	//! First dimension: all fixed-size allocators, second dimension: the buffers of each allocator
	vector<vector<IndexBufferInfo>> buffers;

//! Information to serialize an index buffer to the WAL
struct IndexBufferInfo {
	IndexBufferInfo(data_ptr_t buffer_ptr, const idx_t allocation_size) :
 		buffer_ptr(buffer_ptr), allocation_size(allocation_size) {}

	data_ptr_t buffer_ptr;
	idx_t allocation_size;
};
```

Mytherin commented on src/storage/wal_replay.cpp:
```diff
@@ -403,47 +404,86 @@ void ReplayState::ReplayDropTableMacro(BinaryDeserializer &deserializer) {
-	auto info = deserializer.ReadProperty<unique_ptr<CreateInfo>>(101, "index");
+
+	auto create_info = deserializer.ReadProperty<unique_ptr<CreateInfo>>(101, "index_catalog_entry");
+	auto index_info = deserializer.ReadProperty<IndexStorageInfo>(102, "index_storage_info");
```
Ideally we would avoid calling `ReadProperty` explicitly here and we would move this into a struct whose serialization methods can be generated through the serialize script

Mytherin reviewed (commented):
Thanks for the PR! Looks great - some comments from my side:

taniabogatsch committed:
Merge branch 'feature' into art-wal

taniabogatsch committed:
fix index unique names

taniabogatsch committed:
review feedback

taniabogatsch committed:
missing include

taniabogatsch committed:
more review feedback

taniabogatsch committed:
revert to index_name

taniabogatsch committed:
Merge branch 'main' into art-wal

Mytherin commented:
Thanks!
